File: frontend/src/types/game.ts
/**
 * Game types for Hanyang: The Foundation
 */

// Enums
export type ResourceType = 'wood' | 'stone' | 'tile' | 'ink'
export type WorkerType = 'apprentice' | 'official'
export type GameStatus = 'waiting' | 'in_progress' | 'finished'
export type PlayerColor = 'blue' | 'red' | 'green' | 'yellow'
export type TerrainType = 'normal' | 'mountain' | 'water'
export type ActionType = 'place_worker' | 'recall_worker' | 'place_tile' | 'select_blueprint' | 'end_turn' | 'pass'
export type BlueprintCategory = 'palace_proximity' | 'category_collection' | 'pattern' | 'special'

// Resource & Worker
export interface Resources {
  wood: number
  stone: number
  tile: number
  ink: number
}

export interface WorkerState {
  total: number
  available: number
  placed: number
}

export interface PlayerWorkers {
  apprentices: WorkerState
  officials: WorkerState
}

// Board
export interface BoardPosition {
  row: number
  col: number
}

export interface PlacedWorker {
  player_id: number
  worker_type: WorkerType
  slot_index: number
}

export interface PlacedTile {
  tile_id: string
  owner_id: number
  placed_workers: PlacedWorker[]
  fengshui_active: boolean
}

export interface BoardCell {
  position: BoardPosition
  terrain: TerrainType
  tile: PlacedTile | null
}

// Player
export interface GamePlayer {
  id: number
  user_id: number
  username: string
  color: PlayerColor
  turn_order: number
  is_host: boolean
  is_ready: boolean
  resources: Resources
  workers: PlayerWorkers
  dealt_blueprints: string[]  // 선택 전 배분된 청사진
  blueprints: string[]  // 선택한 청사진
  score: number
  placed_tiles: string[]
}

// Game State
export interface GameState {
  id: number
  lobby_id: number
  status: GameStatus
  current_round: number
  total_rounds: number
  current_turn_player_id: number
  turn_order: number[]
  board: BoardCell[][]
  players: GamePlayer[]
  available_tiles: string[]
  discarded_tiles: string[]
  created_at: string
  updated_at: string
}

// Action Payloads
export interface PlaceWorkerPayload {
  type: 'place_worker'
  worker_type: WorkerType
  target_position: BoardPosition
  slot_index: number
}

export interface RecallWorkerPayload {
  type: 'recall_worker'
  worker_type: WorkerType
  from_position: BoardPosition
  slot_index: number
}

export interface PlaceTilePayload {
  type: 'place_tile'
  tile_id: string
  position: BoardPosition
}

export interface SelectBlueprintPayload {
  type: 'select_blueprint'
  blueprint_id: string
}

export interface EndTurnPayload {
  type: 'end_turn'
}

export type ActionPayload = PlaceWorkerPayload | RecallWorkerPayload | PlaceTilePayload | SelectBlueprintPayload | EndTurnPayload

// Tile Types
export type TileCategory = 'palace' | 'government' | 'religious' | 'commercial' | 'residential' | 'gate'

export interface TileCost {
  wood: number
  stone: number
  tile: number
  ink: number
}

export interface TileInfo {
  tile_id: string
  name_ko: string
  cost: TileCost
  base_points: number
}

export interface TilePlacementAction {
  action_type: 'place_tile'
  available_tiles: TileInfo[]
  valid_positions: BoardPosition[]
}

// Blueprint Types
export interface BlueprintInfo {
  blueprint_id: string
  name_ko: string
  name_en: string
  description_ko: string
  category: BlueprintCategory
  bonus_points: number
  condition: {
    type: string
    params: Record<string, unknown>
  }
  current_score?: number
  is_completed?: boolean
}

export interface BlueprintSelectionAction {
  action_type: 'select_blueprint'
  available_blueprints: BlueprintInfo[]
}

export interface PlayerBlueprintsResponse {
  dealt_blueprints: BlueprintInfo[]
  selected_blueprints: BlueprintInfo[]
}

// API Requests
export interface GameActionRequest {
  action_type: ActionType
  payload: ActionPayload
}

// API Responses
export interface GameActionResponse {
  success: boolean
  action_result: Record<string, unknown>
  new_state: GameState
}

export interface ValidActionOption {
  position?: BoardPosition
  slot_index?: number
}

export interface ValidAction {
  action_type: ActionType
  worker_type?: WorkerType
  available_slots?: ValidActionOption[]
}

export interface ValidActionsResponse {
  valid_actions: ValidAction[]
  message?: string
}

// Score
export interface ScoreBreakdown {
  building_points: number
  fengshui_bonus: number
  adjacency_bonus: number
  blueprint_bonus: number
  remaining_resources: number
  total: number
}

export interface PlayerRanking {
  player_id: number
  username: string
  rank: number
  score_breakdown: ScoreBreakdown
}

export interface GameResultResponse {
  game_id: number
  winner_id: number
  rankings: PlayerRanking[]
  duration_minutes: number
  total_rounds: number
}

// Store State
export interface GameStoreState {
  gameState: GameState | null
  validActions: ValidAction[]
  selectedWorker: WorkerType | null
  selectedTile: string | null
  selectedBlueprint: string | null
  selectedPosition: BoardPosition | null
  playerBlueprints: PlayerBlueprintsResponse | null
  isLoading: boolean
  error: string | null

  // Actions
  fetchGameState: (gameId: number) => Promise<void>
  fetchValidActions: (gameId: number) => Promise<void>
  fetchPlayerBlueprints: (gameId: number) => Promise<void>
  performAction: (gameId: number, action: GameActionRequest) => Promise<void>
  selectWorker: (workerType: WorkerType | null) => void
  selectTile: (tileId: string | null) => void
  selectBlueprint: (blueprintId: string | null) => void
  selectPosition: (position: BoardPosition | null) => void
  clearError: () => void
}

File: backend/app/schemas/game.py
"""
Game schemas - Pydantic models for game state and actions.
Synchronized with contracts/types.ts and contracts/game.contract.ts
"""
from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field


# ============================================
# Enums
# ============================================

class ResourceType(str, Enum):
    WOOD = "wood"
    STONE = "stone"
    TILE = "tile"
    INK = "ink"


class WorkerType(str, Enum):
    APPRENTICE = "apprentice"
    OFFICIAL = "official"


class BuildingCategory(str, Enum):
    PALACE = "palace"
    GOVERNMENT = "government"
    RELIGIOUS = "religious"
    COMMERCIAL = "commercial"
    RESIDENTIAL = "residential"
    GATE = "gate"


class GameStatus(str, Enum):
    WAITING = "waiting"
    IN_PROGRESS = "in_progress"
    FINISHED = "finished"


class PlayerColor(str, Enum):
    BLUE = "blue"
    RED = "red"
    GREEN = "green"
    YELLOW = "yellow"


class ActionType(str, Enum):
    PLACE_WORKER = "place_worker"
    RECALL_WORKER = "recall_worker"
    PLACE_TILE = "place_tile"
    DRAW_BLUEPRINT = "draw_blueprint"
    END_TURN = "end_turn"
    PASS = "pass"


class TerrainType(str, Enum):
    NORMAL = "normal"
    MOUNTAIN = "mountain"
    WATER = "water"


# ============================================
# Resource & Worker
# ============================================

class Resources(BaseModel):
    """자원 보유량"""
    wood: int = 0
    stone: int = 0
    tile: int = 0
    ink: int = 0


class WorkerState(BaseModel):
    """워커 상태"""
    total: int
    available: int
    placed: int = 0


class PlayerWorkers(BaseModel):
    """플레이어 워커 전체 상태"""
    apprentices: WorkerState
    officials: WorkerState


# ============================================
# Board
# ============================================

class BoardPosition(BaseModel):
    """보드 좌표"""
    row: int = Field(..., ge=0, le=4)
    col: int = Field(..., ge=0, le=4)


class PlacedWorker(BaseModel):
    """배치된 워커"""
    player_id: int
    worker_type: WorkerType
    slot_index: int


class PlacedTile(BaseModel):
    """배치된 타일"""
    tile_id: str
    owner_id: int
    placed_workers: list[PlacedWorker] = []
    fengshui_active: bool = False


class BoardCell(BaseModel):
    """보드 셀"""
    position: BoardPosition
    terrain: TerrainType = TerrainType.NORMAL
    tile: PlacedTile | None = None


# ============================================
# Player
# ============================================

class PlayerBase(BaseModel):
    """플레이어 기본 정보"""
    id: int
    user_id: int
    username: str
    color: PlayerColor
    turn_order: int
    is_host: bool = False
    is_ready: bool = False


class GamePlayer(PlayerBase):
    """게임 내 플레이어 상태"""
    resources: Resources = Field(default_factory=Resources)
    workers: PlayerWorkers
    blueprints: list[str] = []
    score: int = 0
    placed_tiles: list[str] = []


# ============================================
# Action Payloads
# ============================================

class PlaceWorkerPayload(BaseModel):
    """워커 배치 페이로드"""
    type: Literal["place_worker"] = "place_worker"
    worker_type: WorkerType
    target_position: BoardPosition
    slot_index: int


class RecallWorkerPayload(BaseModel):
    """워커 회수 페이로드"""
    type: Literal["recall_worker"] = "recall_worker"
    worker_type: WorkerType
    from_position: BoardPosition
    slot_index: int


class PlaceTilePayload(BaseModel):
    """타일 배치 페이로드"""
    type: Literal["place_tile"] = "place_tile"
    tile_id: str
    position: BoardPosition


class DrawBlueprintPayload(BaseModel):
    """청사진 드로우 페이로드"""
    type: Literal["draw_blueprint"] = "draw_blueprint"
    blueprint_id: str


class EndTurnPayload(BaseModel):
    """턴 종료 페이로드"""
    type: Literal["end_turn"] = "end_turn"


class PassPayload(BaseModel):
    """패스 페이로드"""
    type: Literal["pass"] = "pass"


ActionPayload = (
    PlaceWorkerPayload
    | RecallWorkerPayload
    | PlaceTilePayload
    | DrawBlueprintPayload
    | EndTurnPayload
    | PassPayload
)


# ============================================
# Game Action
# ============================================

class GameActionRequest(BaseModel):
    """게임 액션 요청"""
    action_type: ActionType
    payload: ActionPayload


class GameAction(BaseModel):
    """게임 액션"""
    id: int
    game_id: int
    player_id: int
    action_type: ActionType
    payload: dict  # JSON serialized payload
    timestamp: datetime


# ============================================
# Game State
# ============================================

class GameState(BaseModel):
    """게임 전체 상태"""
    id: int
    lobby_id: int
    status: GameStatus
    current_round: int
    total_rounds: int
    current_turn_player_id: int
    turn_order: list[int]
    board: list[list[BoardCell]]  # 5x5 grid
    players: list[GamePlayer]
    available_tiles: list[str]
    discarded_tiles: list[str]
    last_action: GameAction | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ============================================
# Response Types
# ============================================

class BonusAwarded(BaseModel):
    """보너스 정보"""
    fengshui_bonus: int | None = None
    adjacency_bonus: int | None = None
    blueprint_completed: str | None = None


class GameActionResponse(BaseModel):
    """게임 액션 응답"""
    action: GameAction
    new_state: GameState
    bonus_awarded: BonusAwarded | None = None


class ValidActionOption(BaseModel):
    """유효 액션 옵션"""
    position: BoardPosition | None = None
    tile_id: str | None = None
    worker_type: WorkerType | None = None
    slot_index: int | None = None


class ValidAction(BaseModel):
    """유효 액션"""
    action_type: ActionType
    options: list[ValidActionOption] = []


class ValidActionsResponse(BaseModel):
    """유효 액션 목록"""
    valid_actions: list[ValidAction]


class ScoreBreakdown(BaseModel):
    """점수 상세"""
    building_points: int = 0
    fengshui_bonus: int = 0
    adjacency_bonus: int = 0
    blueprint_bonus: int = 0
    remaining_resources: int = 0
    total: int = 0


class PlayerRanking(BaseModel):
    """플레이어 랭킹"""
    player_id: int
    username: str
    rank: int
    score_breakdown: ScoreBreakdown


class GameResultResponse(BaseModel):
    """게임 결과 응답"""
    game_id: int
    winner_id: int
    rankings: list[PlayerRanking]
    duration_minutes: int
    total_rounds: int


class SubmitFeedbackRequest(BaseModel):
    """피드백 제출 요청"""
    rating: int = Field(..., ge=1, le=5)
    comments: str
    balance_feedback: str | None = None
    bug_reports: str | None = None
